### PR TITLE
Hierarchical names

### LINKED ISSUE #23: Allow hierarchical names
Currently, directories in a warehouse have rigidly defined semantics: `<type>/<object name>/`. This is a bit restrictive, since it makes it impossible to have e.g. `service/dhcp-server/lab-network/`, which we probably want. We probably still want the first level to be a type and the entire rest of the path to be collected into an object name.

### PR DESCRIPTION
**Pull request by notCalle**
Reimplementation of the whole internal Pallet loading API to facilitate more advanced naming and filesystem path management. Recursively load file system hierarchical child pallets. Public API as far as current known usage should be intact, except for `PalletJack#pallets` that needed to be removed to keep private data structure changes private.

The old identity query API `Pallet#kind` and `Pallet#name` will return the topmost (still the `#kind`) and bottom-most (the `#leaf_name`) path components for a pallet within the warehouse. This means that a warehouse user can use filesystem hierarchy for grouping of related objects without breaking anything. For true hierarchical use, related tools have to be updated to make use of the new identity API `Pallet#full_name`, `Pallet#leaf_name`, and `Pallet#parent_name`.

Bump version to '0.2.0'.

This closes #23.

### CODE REVIEW

File: lib/palletjack/pallet/identity.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      # Full expanded path to this pallet
+      attr_reader :path
+      
+      # Base path of the warehouse for this pallet
+      attr_reader :warehouse
+      
+      # The kind of this pallet
+      attr_reader :kind
+      
+      # The hierarchical _parent name for this pallet, or nil
```
Comment (creideiki) [inline lib/palletjack/pallet/identity.rb]: I don't understand what this comment is trying to say.

**Review by creideiki — CHANGES_REQUESTED** [review]

File: lib/palletjack/pallet.rb [hunk 2]
```diff
@@ -65,10 +61,39 @@ def inspect
...
+
+    alias name leaf_name
+
+    # The full hierarchical name of this pallet
+
+    def full_name
+      @identity.full_name
+    end
+
+    # The hierarchical _parent name for this pallet, or nil
```
Comment (creideiki) [inline lib/palletjack/pallet.rb]: As in `identity.rb` earlier, this comment is not clear.

**Review by creideiki — CHANGES_REQUESTED** [review]

**Comment by creideiki** [comment]
`dump_pallet` should be updated as well, to accept full path names. Currently, to get the DNS resolver data out of the warehouse, you need to run:

```
dump_pallet -w ~/src/palletjack/examples/warehouse/ -t service example-com
```

which doesn't scale. `dump_pallet` should have some way of accepting a fully qualified name, e.g.`dns-resolver/example-com`.

**Comment by notCalle** [comment]
Rebased, because `dump_pallet` conflicted.

**Pull request merged**